crates/kornia-imgproc/benches/bench_gradient.rs

```diff
@@ -0,0 +1,600 @@
+use criterion::{black_box, criterion_group, criterion_main, BenchmarkId, Criterion};
+
+use kornia_image::{Image, ImageError};
+use kornia_imgproc::filter::{
+    kernels, spatial_gradient_float, spatial_gradient_float_by_separable_filter,
+    spatial_gradient_float_parallel,
+};
+use rayon::{
+    iter::{IndexedParallelIterator, ParallelIterator},
+    slice::ParallelSliceMut,
+};
+
+/// Compute the first order image derivative in both x and y using a Sobel operator.
+///
+/// # Arguments
+///
+/// * `src` - The source image with shape (H, W).
+/// * `dst` - The destination image with shape (H, W, 2).
+pub fn spatial_gradient_float_rayon_parallel<const C: usize>(
+    src: &Image<f32, C>,
+    dx: &mut Image<f32, C>,
+    dy: &mut Image<f32, C>,
+) -> Result<(), ImageError> {
+    if src.size() != dx.size() {
+        return Err(ImageError::InvalidImageSize(
+            src.cols(),
+            src.rows(),
+            dx.cols(),
+            dx.rows(),
+        ));
+    }
+
+    if src.size() != dy.size() {
+        return Err(ImageError::InvalidImageSize(
+            src.cols(),
+            src.rows(),
+            dy.cols(),
+            dy.rows(),
+        ));
+    }
+
+    let (sobel_x, sobel_y) = kernels::normalized_sobel_kernel3();
+    let cols = src.cols();
+
+    let src_data = src.as_slice();
+
+    dx.as_slice_mut()
+        .par_chunks_mut(cols * C)
+        .zip(dy.as_slice_mut().par_chunks_mut(cols * C))
+        .enumerate()
+        .for_each(|(r, (dx_row, dy_row))| {
+            dx_row
+                .par_chunks_mut(C)
+                .zip(dy_row.par_chunks_mut(C))
+                .enumerate()
+                .for_each(|(c, (dx_c, dy_c))| {
+                    for ch in 0..C {
+                        let mut sum_x = 0.0;
+                        let mut sum_y = 0.0;
+                        for dy in 0..3 {
+                            for dx in 0..3 {
+                                let row = (r + dy).min(src.rows()).max(1) - 1;
+                                let col = (c + dx).min(src.cols()).max(1) - 1;
+                                let src_pix_offset = (row * src.cols() + col) * C + ch;
+                                let val = unsafe { src_data.get_unchecked(src_pix_offset) };
+                                sum_x += val * sobel_x[dy][dx];
+                                sum_y += val * sobel_y[dy][dx];
+                            }
+                        }
+                        dx_c[ch] = sum_x;
+                        dy_c[ch] = sum_y;
+                    }
+                });
+        });
+
+    Ok(())
+}
+
+/// Compute the first order image derivative in both x and y using a Sobel operator.
+///
+/// # Arguments
+///
+/// * `src` - The source image with shape (H, W).
+/// * `dst` - The destination image with shape (H, W, 2).
+pub fn spatial_gradient_float_rayon_row_parallel<const C: usize>(
```

Comment:
let;s choose this one but there's too much boilerplate code in the branching

Reply:
Code is ugly now for easily developing, I will clean it up as final code